README.md
Node.js - github-download
================

[![build status](https://secure.travis-ci.org/jprichardson/node-github-download.svg)](http://travis-ci.org/jprichardson/node-github-download)

Easily download Github repos without any external dependencies such as Git, Tar, Unzip, etc.


Why?
----

I really like the concept of managing user defined projects, repos, file structures (package management) on Github like the way that [Component](https://github.com/component) does package management. I have a package management system [Rock](https://github.com/rocktemplates) that I use to create skeleton/templates of projects. I wanted Rock to use Github as a package management system. I also didn't want any dependencies amongst any 3rd party programs like Git, Tar, or Unzip. Pure Node.js JavaScript is what I wanted.


Installation
------------

    npm install github-download



Usage
-----

### ghdownload(params, dir)

Downloads the latest copy of some Github reference (branch, tag, or commit), or the `master` branch by default (specifically the `master` branch, it does _not_ honor Github's default branch configuration). This will still work even if the Github API limit has been reached.

- **params**: Can either be:
     - a Github URL string such as:
         - `https://github.com/jprichardson/node-vcsurl.git`
         - `git@github.com:jprichardson/node-vcsurl.git`
         - `git://github.com/jprichardson/node-vcsurl.git`
         - and even including a reference, e.g. `https://github.com/jprichardson/node-vcsurl.git#master`
     - or an object like so: `{user: 'jprichardson', repo: 'vcsurl', ref: 'master'}`
- **dir**: The output directory. Uses the current working directory if nothing is specified.

Returns a GithubDownloader object that emits events on `dir`, `file`, and `end`.

Example:

```javascript
var ghdownload = require('github-download')
  , exec = require('exec')

ghdownload({user: 'jprichardson', repo: 'node-batchflow', ref: 'master'}, process.cwd())
.on('dir', function(dir) {
  console.log(dir)
})
.on('file', function(file) {
  console.log(file)
})
.on('zip', function(zipUrl) { //only emitted if Github API limit is reached and the zip file is downloaded
  console.log(zipUrl)
})
.on('error', function(err) {
  console.error(err)
})
.on('end', function() {
  exec('tree', function(err, stdout, sderr) {
    console.log(stdout)
  })
})
```

Outputs:

    .
    ├── CHANGELOG.md
    ├── LICENSE
    ├── README.md
    ├── lib
    │   └── batchflow.js
    ├── package.json
    └── test
        ├── batchflow-par-array.test.js
        ├── batchflow-par-limit.test.js
        ├── batchflow-par-object.test.js
        ├── batchflow-seq-array.test.js
        ├── batchflow-seq-object.test.js
        ├── batchflow.test.js
        ├── mocha.opts
        └── resources

    3 directories, 12 files



License
-------

(MIT License)

Copyright 2013-2016, JP Richardson  <jprichardson@gmail.com>


CHANGELOG.md
0.5.0 / 2016-05-04
------------------
- standard update
- fix proper path when downloading ref [#14][#14]

0.4.0 / 2015-08-30
------------------
- JavaScript Standard Style http://standardjs.com
- upgraded to `fs-extra@0.24.x`

0.3.0 / 2014-06-17
------------------
* downloads to a temp dir in `cwd()` instead of `/tmp` [#4](https://github.com/jprichardson/node-github-download/pull/4)
* upgraded `"fs-extra": "~0.6.0"` to `"fs-extra": "^0.9.1"`
* drop support for Node.js `v0.8`

0.2.0 / 2013-04-29
------------------
* updated `fs-extra` dep
* Node `v0.10` compatibility
* fixed inheritance bug
* add travis-ci

0.1.1 / 2013-02-03
------------------
* Updated `fs-extra` dep.

0.1.0 / 2013-01-21
------------------
* Updated dependencies to include `adm-zip` and `fs-extra`.
* If Github API limit is hit (it's pretty low), then it'll download the zip file and extract the contents.

0.0.1 / 2013-01-17
------------------
* Initial release.

[#14]: https://github.com/jprichardson/node-github-download/pull/14      "Fix downloading specific version"
[#13]: https://github.com/jprichardson/node-github-download/issues/13    "Retrieve commit hash"
[#12]: https://github.com/jprichardson/node-github-download/issues/12    "errno -4048 rename when download"
[#11]: https://github.com/jprichardson/node-github-download/issues/11    "downloading parts of a repo"
[#10]: https://github.com/jprichardson/node-github-download/issues/10    "Update to latest version of fs-extra"
[#9]: https://github.com/jprichardson/node-github-download/issues/9      "Downloading from private repos"
[#8]: https://github.com/jprichardson/node-github-download/issues/8      "Pointing at a tag"
[#7]: https://github.com/jprichardson/node-github-download/issues/7      "github-download is not working"
[#6]: https://github.com/jprichardson/node-github-download/pull/6        "bug-fix: download files across devices now works"
[#5]: https://github.com/jprichardson/node-github-download/issues/5      "Make it possible to target a release"
[#4]: https://github.com/jprichardson/node-github-download/pull/4        "Downloading onto a different drive throws an EXDEV error"
[#3]: https://github.com/jprichardson/node-github-download/pull/3        "Added support for GitHub refs (branches, tags, commits)"
[#2]: https://github.com/jprichardson/node-github-download/issues/2      "Upgrade to working adm-zip [backlog]"
[#1]: https://github.com/jprichardson/node-github-download/issues/1      "Delete downloaded zip file [backlog]"


test/VERIFY.js
var path = require('path')
  , fs = require('fs-extra')
  , P = require('autoresolve')
  , ghdownload = require(P('lib/github-download'))


module.exports = VERIFY

function VERIFY (input, outputDir, transportConfig, done) {
  var results = {}

  if (!fs.existsSync(outputDir)) {
    fs.removeSync(outputDir)
    fs.mkdirsSync(outputDir)
  }

  ghdownload(input, outputDir, transportConfig)
  .on('dir', function(dir) {
    if (!results['dir']) results['dir'] = []
    results['dir'].push(dir)
  })
  .on('file', function(file) {
    if (!results['file']) results['file'] = []
    results['file'].push(file)
  })
  .on('zip', function(zipUrl) {
    results['zip'] = zipUrl
  })
  .on('git', function(gitUrl) {
    results['git'] = gitUrl
  })
  .on('error', function(err) {
    if (!results['error']) results['error'] = []
    results['error'].push(err)
  })
  .on('end', function(success) {
    results['success'] = success

    var outputFiles = []

    outputFiles.push(path.join(outputDir, '.gitignore'))
    outputFiles.push(path.join(outputDir, '.travis.yml'))
    outputFiles.push(path.join(outputDir, 'CHANGELOG.md'))
    outputFiles.push(path.join(outputDir, 'LICENSE'))
    outputFiles.push(path.join(outputDir, 'README.md'))
    outputFiles.push(path.join(outputDir, 'package.json'))
    outputFiles.push(path.join(outputDir, 'lib/batchflow.js'))
    outputFiles.push(path.join(outputDir, 'test/batchflow.test.js'))
    outputFiles.push(path.join(outputDir, 'test/mocha.opts'))

    results['exist'] = {}

    outputFiles.forEach(function(file) {
      results['exist'][file] = fs.existsSync(file)
    })

    //has some correct content
    var licenseFile = path.join(outputDir,'LICENSE')
    results['content'] = fs.existsSync(licenseFile)
    if (results['content'])
      results['content'] = fs.readFileSync(licenseFile, 'utf8').indexOf('jprichardson@gmail.com') > 0
    
    setTimeout(function() {
      done(results) 
    },250)
  })
}



test/transport_api.test.js
var ghdownload = require('../lib/github-download')
  , testutil = require('testutil')
  , fs =require('fs-extra')
  , path = require('path')
  , nock = require('nock')
  , P = require('autoresolve')
  , VERIFY = require(P('test/VERIFY'))

var TEST_DIR = null

describe('github-download / api transport', function() {
  beforeEach(function() {
    TEST_DIR = testutil.createTestDir('github-download')
    TEST_DIR = path.join(TEST_DIR, 'transport-api')
    fs.mkdirsSync(TEST_DIR)
  })

  describe('> when input is an object with user and repo', function() {
    it('should download the latest copy of master', function(done) {
      var transportConfig = {order: ['api']}
      var input = {user: 'jprichardson', repo: 'node-batchflow'}
      VERIFY(input, TEST_DIR, transportConfig, function(results) {
        F (results['zip'])
        F (results['git'])
        T (results['dir'])
        T (results['file'])
        F (results['error'])
        T (results['success'])

        var filesExist = results['exist']
        T (Object.keys(filesExist).every(function(file) { return filesExist[file] }))
        T (results['content'])
        done()
      })
    })
  })

  describe('> when input is an relative Github repo', function() {
    it('should download the latest copy of master', function(done) {
      var transportConfig = {order: ['api']}
      var input = 'jprichardson/node-batchflow'
      VERIFY(input, TEST_DIR, transportConfig, function(results) {
        F (results['zip'])
        F (results['git'])
        T (results['dir'])
        T (results['file'])
        F (results['error'])
        T (results['success'])

        var filesExist = results['exist']
        T (Object.keys(filesExist).every(function(file) { return filesExist[file] }))
        T (results['content'])
        done()
      })
    })
  })

  describe('> when input is git path Github repo', function() {
    it('should download the latest copy of master', function(done) {
      var transportConfig = {order: ['api']}
      var input = 'git@github.com:jprichardson/node-batchflow.git'
      VERIFY(input, TEST_DIR, transportConfig, function(results) {
        F (results['zip'])
        F (results['git'])
        T (results['dir'])
        T (results['file'])
        F (results['error'])
        T (results['success'])

        var filesExist = results['exist']
        T (Object.keys(filesExist).every(function(file) { return filesExist[file] }))
        T (results['content'])
        done()
      })
    })
  })

  describe('> when Github API limit has been reached', function() {
    it('should download the zip of the repo', function(done) {
      var transportConfig = {order: ['api']}
      var input = {user: 'jprichardson', repo: 'node-batchflow'}
      
      var scope = nock('https://api.github.com/')
      .filteringPath(function(path) {
        return '*';
      })
      .get('*')
      .reply(403, {message: "API Rate Limit Exceeded for $IP"})

      VERIFY(input, TEST_DIR, transportConfig, function(results) {
        F (results['zip'])
        F (results['git'])
        F (results['dir'])
        F (results['file'])
        T (results['error'])
        F (results['success'])

        var filesExist = results['exist']
        T (Object.keys(filesExist).every(function(file) { return !filesExist[file] }))
        F (results['content'])
        done()
      })
    })
  })
})



test/github-download.test.js
var ghdownload = require('../lib/github-download')
var testutil = require('testutil')
var fs = require('fs-extra')
var path = require('path')
var nock = require('nock')

var TEST_DIR = null

/* global beforeEach describe it T */

describe('github-download', function () {
  beforeEach(function () {
    if (fs.existsSync(TEST_DIR)) fs.removeSync(TEST_DIR)
    TEST_DIR = testutil.createTestDir('github-download')
  })

  describe('> when input is an object with user and repo', function () {
    it('should download the latest copy of master', function (done) {
      var input = {user: 'jprichardson', repo: 'node-batchflow'}
      TEST(path.join(TEST_DIR, '1'), input, done)
    })
  })

  describe('> when input is an relative Github repo', function () {
    it('should download the latest copy of master', function (done) {
      var input = 'jprichardson/node-batchflow'
      TEST(path.join(TEST_DIR, '2'), input, done)
    })
  })

  describe('> when input is git path Github repo', function () {
    it('should download the latest copy of master', function (done) {
      var input = 'git@github.com:jprichardson/node-batchflow.git'
      TEST(path.join(TEST_DIR, '3'), input, done)
    })
  })

  describe('> when Github API limit has been reached', function () {
    it('should download the zip of the repo', function (done) {
      var input = {user: 'jprichardson', repo: 'node-batchflow'}

      nock('https://api.github.com/')
        .filteringPath(function (path) {
          return '*'
        })
        .get('*')
        .reply(403, {message: 'API Rate Limit Exceeded for $IP'})

      TEST(path.join(TEST_DIR, '4'), input, done)
    })
  })
})

function TEST (outputDir, input, done) {
  var files = []
  var dirs = []
  var zip = null

  // console.log('OD: ' + outputDir)
  if (!fs.existsSync(outputDir)) {
    fs.removeSync(outputDir)
    fs.mkdirsSync(outputDir)
  }

  ghdownload(input, outputDir)
    .on('dir', function (dir) {
      // console.log('D: ' + dir)
      dirs.push(dir)
    })
    .on('file', function (file) {
      // console.log('F: ' + file)
      files.push(file)
    })
    .on('zip', function (zipUrl) {
      zip = zipUrl
    })
    .on('error', function (err) {
      done(err) // shouldn't happen
    })
    .on('end', function () {
      if (!zip) { // file and dir events dont get emitted on zip
        T(files.length > 10)
        T(dirs.length >= 2)
      }

      var items = []

      // console.log('OD2: ' + outputDir)
      items = items.concat(fs.readdirSync(outputDir))
      items = items.concat(fs.readdirSync(path.join(outputDir, 'lib')))
      items = items.concat(fs.readdirSync(path.join(outputDir, 'test')))
      // console.dir(items)

      T(fs.existsSync(path.join(outputDir, '.gitignore')))
      T(fs.existsSync(path.join(outputDir, '.travis.yml')))
      T(fs.existsSync(path.join(outputDir, 'CHANGELOG.md')))
      T(fs.existsSync(path.join(outputDir, 'LICENSE')))
      T(fs.existsSync(path.join(outputDir, 'README.md')))
      T(fs.existsSync(path.join(outputDir, 'package.json')))
      T(fs.existsSync(path.join(outputDir, 'lib/batchflow.js')))
      T(fs.existsSync(path.join(outputDir, 'test/batchflow.test.js')))
      T(fs.existsSync(path.join(outputDir, 'test/mocha.opts')))

      T(fs.readFileSync(path.join(outputDir, 'LICENSE'), 'utf8').indexOf('jprichardson@gmail.com') > 0)

      setTimeout(function () {
        done()
      }, 20)
    })
}


lib/transport_api.js
var EventEmitter = require('events').EventEmitter
  , vcsurl = require('vcsurl')
  , request = require('request')
  , path = require('path')
  , fs = require('fs-extra')
  , os = require('os')
  
module.exports.GithubApiDownloader = GithubApiDownloader

function GithubApiDownloader (user, repo, branch, outputDir, options) {
  this.user = user
  this.repo = repo
  this.branch = branch || 'master'
  this.dir = outputDir
  this._log = []
  this._hasError = false
  this.options = options || {}
  this.on('error', function() {
    this._hasError = true
  })
}
GithubApiDownloader.prototype = new EventEmitter();

GithubApiDownloader.prototype.start = function() {
  var _this = this 
    , initialUrl = 'https://api.github.com/repos/' + this.user + '/' + this.repo + '/contents/'
    , rawUrl = 'https://raw.github.com/' + this.user + '/' + this.repo + '/' + this.branch + '/'
    , pending = 0
    , gonnaProcess = 0

  gonnaProcess += 1
  requestJSON.call(this, initialUrl, processItems)

  function processItems (items) {
    gonnaProcess -= 1

    if (items && items.length > 0) {
      pending += items.length
      items.forEach(handleItem)
    }
    checkDone()
  }

  function handleItem (item) {
    if (item.type === 'dir') {
      var dir = path.join(_this.dir, item.path)
      fs.mkdirs(dir, function (err) {
        if (err) _this.emit('error', err)
        _this._log.push(dir)
        gonnaProcess += 1
        requestJSON.call(_this, initialUrl + item.path, processItems)
        _this.emit('dir', item.path)
        pending -= 1
        checkDone()
      })
    } else if (item.type === 'file') {
      var file = path.join(_this.dir, item.path)
      fs.createFile(file, function(err) {
        if (err) _this.emit('error', err)
        request.get(rawUrl + item.path).pipe(fs.createWriteStream(file)).on('close', function() {
          _this._log.push(file)
          _this.emit('file', item.path)
          pending -= 1
          checkDone()
        })
      })
    }
    else
      _this.emit('Error', new Error(JSON.stringify(item, null, 2) + '\n does not have type.'))
  }

  function checkDone () {
    //console.log('PENDING: ' + pending + ' gonnaProcess: ' + gonnaProcess)
    if (pending === 0 && gonnaProcess === 0) {
      _this.emit('end', !_this._hasError)
    }
  }

  return this;
}



/****************************
 * PRIVATE METHODS
 ****************************/

function requestJSON (url, callback) {
  var _this = this
  request(, function(err, resp, body) {
    if (err) return this.emit('error', err)
    if (resp.statusCode !== 200) _this.emit('error', new Error(url + ': returned ' + resp.statusCode + '\n\nbody:\n' + body))

    callback(JSON.parse(body))
 })
}










lib/github-download.js
var EventEmitter = require('events').EventEmitter
var vcsurl = require('vcsurl')
var request = require('request')
var path = require('path')
var fs = require('fs-extra')
var util = require('util')
var extractZip = require('./zip').extractZip
var cwd = process.cwd()

function GithubDownloader (user, repo, ref, dir) {
  this.user = user
  this.repo = repo
  this.ref = ref || 'master'
  this.dir = dir
  this._log = []
  this._getZip = false
}
util.inherits(GithubDownloader, EventEmitter)

GithubDownloader.prototype.start = function () {
  var _this = this
  var initialUrl = 'https://api.github.com/repos/' + this.user + '/' + this.repo + '/contents/'
  var initialUrlRef = this.ref ? '?ref=' + this.ref : ''
  var rawUrl = 'https://raw.github.com/' + this.user + '/' + this.repo + '/' + this.ref + '/'
  var pending = 0
  var gonnaProcess = 0

  gonnaProcess += 1
  requestJSON.call(this, initialUrl + initialUrlRef, processItems)

  function processItems (items) {
    pending += items.length
    gonnaProcess -= 1
    items.forEach(handleItem)
    checkDone()
  }

  function handleItem (item) {
    if (item.type === 'dir') {
      var dir = path.join(_this.dir, item.path)
      fs.mkdirs(dir, function (err) {
        if (err) _this.emit('error', err)
        _this._log.push(dir)
        gonnaProcess += 1
        requestJSON.call(_this, initialUrl + item.path + initialUrlRef, processItems)
        _this.emit('dir', item.path)
        pending -= 1
        checkDone()
      })
    } else if (item.type === 'file') {
      var file = path.join(_this.dir, item.path)
      fs.createFile(file, function (err) {
        if (err) _this.emit('error', err)
        request.get(rawUrl + item.path).pipe(fs.createWriteStream(file)).on('close', function () {
          _this._log.push(file)
          _this.emit('file', item.path)
          pending -= 1
          checkDone()
        })
      })
    } else {
      _this.emit('Error', new Error(JSON.stringify(item, null, 2) + '\n does not have type.'))
    }
  }

  function checkDone () {
    // console.log('PENDING: ' + pending + ' gonnaProcess: ' + gonnaProcess)
    if (pending === 0 && gonnaProcess === 0 && !this._getZip) {
      _this.emit('end')
    }
  }

  return this
}

module.exports = function GithubDownload (params, dir) {
  if (typeof params === 'string') {
    var pieces = params.split('#')
    var ref = pieces[1]
    var url = (vcsurl(pieces[0]) || pieces[0]).split('/')
    params = {user: url[url.length - 2], repo: url[url.length - 1], ref: ref}
  }

  if (typeof params !== 'object') {
    throw new Error('Invalid parameter type. Should be repo URL string or object containing repo and user.')
  }

  // console.dir(params)

  dir = dir || process.cwd()
  var gh = new GithubDownloader(params.user, params.repo, params.ref, dir)
  return gh.start()
}

// PRIVATE METHODS

function requestJSON (url, callback) {
  var _this = this
  request({url: url}, function (err, resp, body) {
    if (err) return this.emit('error', err)
    if (resp.statusCode === 403) return downloadZip.call(_this)
    if (resp.statusCode !== 200) _this.emit('error', new Error(url + ': returned ' + resp.statusCode + '\n\nbody:\n' + body))

    callback(JSON.parse(body))
  })
}

function downloadZip () {
  var _this = this
  if (_this._getZip) return
  _this._getZip = true

  _this._log.forEach(function (file) {
    fs.remove(file)
  })

  var tmpdir = generateTempDir()
  var zipBaseDir = _this.repo + '-' + _this.ref
  var zipFile = path.join(tmpdir, zipBaseDir + '.zip')

  var zipUrl = 'https://nodeload.github.com/' + _this.user + '/' + _this.repo + '/zip/' + _this.ref
  _this.emit('zip', zipUrl)

  // console.log(zipUrl)
  fs.mkdir(tmpdir, function (err) {
    if (err) _this.emit('error', err)
    request.get(zipUrl).pipe(fs.createWriteStream(zipFile)).on('close', function () {
      extractZip.call(_this, zipFile, tmpdir, function (extractedFolderName) {
        var oldPath = path.join(tmpdir, extractedFolderName)
        fs.rename(oldPath, _this.dir, function (err) {
          if (err) _this.emit('error', err)
          fs.remove(tmpdir, function (err) {
            if (err) _this.emit('error', err)
            _this.emit('end')
          })
        })
      })
    })
  })
}

function generateTempDir () {
  return path.join(cwd, Date.now().toString() + '-' + Math.random().toString().substring(2))
}
